ccl/sqlproxyccl: TestCancelQuery failed

ccl/sqlproxyccl.TestCancelQuery [failed](https://mesolite.cluster.engflow.com/invocations/default/fbce7f04-dc49-4c87-ac1d-5206d445513e?testReportRun=1&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NxbHByb3h5Y2NsOnNxbHByb3h5Y2NsX3Rlc3Q=) on release-24.3.20-rc @ [917f8feb69e137770694819f17dc31363aa82ec7](https://github.com/cockroachdb/cockroach/commits/917f8feb69e137770694819f17dc31363aa82ec7):

Fatal error:

```
panic: use of Span after Finish. Span: proxy-con-serve. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: proxy-con-serve. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 4767 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc004c68c70?, {0xe3c0518, 0xc00d67a180})
	pkg/util/stop/stopper.go:226 +0x74
panic({0xa7611a0?, 0xc0020fb5a0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc0071b2008)
	pkg/util/tracing/span.go:177 +0x274
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc0071b2008)
	pkg/util/tracing/span.go:437 +0x27
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	pkg/util/tracing/span.go:445
github.com/cockroachdb/cockroach/pkg/util/log.getSpan({0xe3c03c0, 0xc0072b1500})
	pkg/util/log/trace.go:22 +0x45
github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0xe3c03c0, 0xc0072b1500}, 0x2, 0x3, 0x1, 0x0, {0xaf57945, 0x27}, {0xc00909b698, 0x2, ...})
	pkg/util/log/channels.go:99 +0x66b
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
	pkg/util/log/channels.go:34
github.com/cockroachdb/cockroach/pkg/util/log.loggerOps.Errorf(...)
	bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:1281
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*connector).dialTenantCluster(0xc008005100, {0xe3c03c0, 0xc0072b1500}, {0xe3c0270, 0xc00448b900})
	pkg/ccl/sqlproxyccl/connector.go:243 +0xcd8
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*connector).OpenTenantConnWithToken(0xc008005100, {0xe3c03c0, 0xc0072b1500}, {0xe3c0270, 0xc00448b900}, {0xc003e5a460, 0x9c})
	pkg/ccl/sqlproxyccl/connector.go:106 +0x1ba
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.transferConnection(0xc0072b1500, {0xe3c0270, 0xc00448b900}, 0xc008005100, 0xc004fd37a0, 0xc006d7c5d0, 0xc00783e750)
	pkg/ccl/sqlproxyccl/conn_migration.go:274 +0x40d
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).TransferConnection(0xc00448b900)
	pkg/ccl/sqlproxyccl/conn_migration.go:213 +0x509
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/balancer.(*Balancer).processQueue.func1.1({0xe3c0518, 0xc00d67a180})
	pkg/ccl/sqlproxyccl/balancer/balancer.go:371 +0x268
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0xae98922, 0x11})
	pkg/util/stop/stopper.go:498 +0x339
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 4545
	pkg/util/stop/stopper.go:488 +0x786
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestCancelQuery
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestCancelQuery3798647868
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestCancelQuery/cancel_after_migrating_a_session
=== RUN   TestCancelQuery/cancel_over_http
=== RUN   TestCancelQuery/cancel_over_sql
=== RUN   TestCancelQuery/forward_over_http
=== RUN   TestCancelQuery/ignore_unknown_secret_key
*
* ERROR: a panic has occurred!
* use of Span after Finish. Span: proxy-con-serve. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | [...repeated from below...]
* Wraps: (2) assertion failure
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
*   | 	pkg/util/tracing/span.go:177
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType
*   | 	pkg/util/tracing/span.go:437
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose
*   | 	pkg/util/tracing/span.go:445
*   | github.com/cockroachdb/cockroach/pkg/util/log.getSpan
*   | 	pkg/util/log/trace.go:22
*   | github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal
*   | 	pkg/util/log/channels.go:99
*   | github.com/cockroachdb/cockroach/pkg/util/log.logfDepth
*   | 	pkg/util/log/channels.go:34
*   | github.com/cockroachdb/cockroach/pkg/util/log.loggerOps.Errorf
*   | 	bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:1281
*   | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*connector).dialTenantCluster
*   | 	pkg/ccl/sqlproxyccl/connector.go:243
*   | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*connector).OpenTenantConnWithToken
*   | 	pkg/ccl/sqlproxyccl/connector.go:106
*   | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.transferConnection
*   | 	pkg/ccl/sqlproxyccl/conn_migration.go:274
*   | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).TransferConnection
*   | 	pkg/ccl/sqlproxyccl/conn_migration.go:213
*   | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/balancer.(*Balancer).processQueue.func1.1
*   | 	pkg/ccl/sqlproxyccl/balancer/balancer.go:371
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (4) use of Span after Finish. Span: proxy-con-serve. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #151664 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot T-serverless branch-release-24.3.19-rc]
- #151651 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot T-serverless branch-release-24.1.23-rc]
- #149408 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot P-3 T-cloud-platform branch-release-25.3]
- #148599 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot T-serverless branch-release-24.1]
- #145279 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot T-serverless branch-release-25.1]
- #144691 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot T-serverless branch-release-24.3]
- #142810 ccl/sqlproxyccl: TestCancelQuery failed [C-test-failure O-robot T-serverless branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCancelQuery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54499

--- a/pkg/ccl/sqlproxyccl/balancer/balancer.go
+++ b/pkg/ccl/sqlproxyccl/balancer/balancer.go
@@ -371,7 +371,7 @@ func (b *Balancer) processQueue(ctx context.Context) {
 
 			// Each request is retried up to maxTransferAttempts.
 			for i := 0; i < maxTransferAttempts && ctx.Err() == nil; i++ {
-				err := req.conn.TransferConnection()
+				err := req.conn.TransferConnection(ctx)
 				if err == nil || errors.Is(err, context.Canceled) {
 					break
 				}

--- a/pkg/ccl/sqlproxyccl/balancer/conn.go
+++ b/pkg/ccl/sqlproxyccl/balancer/conn.go
@@ -19,7 +19,7 @@ type ConnectionHandle interface {
 	// TransferConnection performs a connection migration on the connection
 	// handle. Invoking this blocks until the connection migration process has
 	// been completed.
-	TransferConnection() error
+	TransferConnection(context.Context) error
 
 	// IsIdle returns true if the connection is idle, and false otherwise.
 	IsIdle() bool

--- a/pkg/ccl/sqlproxyccl/balancer/conn_test.go
+++ b/pkg/ccl/sqlproxyccl/balancer/conn_test.go
@@ -44,7 +44,7 @@ func (h *testConnHandle) Close() {
 }
 
 // TransferConnection implements the ConnectionHandle interface.
-func (h *testConnHandle) TransferConnection() error {
+func (h *testConnHandle) TransferConnection(ctx context.Context) error {
 	h.mu.Lock()
 	defer h.mu.Unlock()
 	h.mu.onTransferConnectionCount++

--- a/pkg/ccl/sqlproxyccl/conn_migration.go
+++ b/pkg/ccl/sqlproxyccl/conn_migration.go
@@ -127,7 +127,7 @@ var errTransferCannotStart = errors.New("transfer cannot be started")
 // error).
 //
 // TransferConnection implements the balancer.ConnectionHandle interface.
-func (f *forwarder) TransferConnection() (retErr error) {
+func (f *forwarder) TransferConnection(ctx context.Context) (retErr error) {
 	// A previous non-recoverable transfer would have closed the forwarder, so
 	// return right away.
 	if f.ctx.Err() != nil {
@@ -146,18 +146,18 @@ func (f *forwarder) TransferConnection() (retErr error) {
 	// whenever the context expires. We have to close the forwarder because
 	// the transfer may be blocked on I/O, and the only way for now is to close
 	// the connections. This then allow TransferConnection to return and cleanup.
-	ctx, cancel := newTransferContext(f.ctx)
+	transferCtx, cancel := newTransferContext(ctx)
 	defer cancel()
 
 	// Use a separate handler for timeouts. This is the only way to handle
 	// blocked I/Os as described above.
 	go func() {
-		<-ctx.Done()
+		<-transferCtx.Done()
 		// This Close call here in addition to the one in the defer callback
 		// below is on purpose. This would help unblock situations where we're
 		// blocked on sending/reading messages from connections that couldn't
 		// be handled with context.Context.
-		if !ctx.isRecoverable() {
+		if !transferCtx.isRecoverable() {
 			f.Close()
 		}
 	}()
@@ -179,7 +179,7 @@ func (f *forwarder) TransferConnection() (retErr error) {
 
 		// When TransferConnection returns, it's either the forwarder has been
 		// closed, or the procesors have been resumed.
-		if !ctx.isRecoverable() {
+		if !transferCtx.isRecoverable() {
 			log.Dev.Infof(logCtx, "transfer failed: connection closed, latency=%v, err=%v", latencyDur, retErr)
 			f.metrics.ConnMigrationErrorFatalCount.Inc(1)
 			f.Close()
@@ -201,16 +201,16 @@ func (f *forwarder) TransferConnection() (retErr error) {
 
 	// Suspend both processors before starting the transfer.
 	request, response := f.getProcessors()
-	if err := request.suspend(ctx); err != nil {
+	if err := request.suspend(transferCtx); err != nil {
 		return errors.Wrap(err, "suspending request processor")
 	}
-	if err := response.suspend(ctx); err != nil {
+	if err := response.suspend(transferCtx); err != nil {
 		return errors.Wrap(err, "suspending response processor")
 	}
 
 	// Transfer the connection.
 	clientConn, serverConn := f.getConns()
-	newServerConn, err := transferConnection(ctx, f, f.connector, f.metrics, clientConn, serverConn)
+	newServerConn, err := transferConnection(transferCtx, f, f.connector, f.metrics, clientConn, serverConn)
 	if err != nil {
 		return errors.Wrap(err, "transferring connection")
 	}

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -2018,12 +2018,12 @@ func TestConnectionMigration(t *testing.T) {
 		require.Equal(t, totalAttempts, count)
 	}
 
-	transferConnWithRetries := func(t *testing.T, f *forwarder) error {
+	transferConnWithRetries := func(t *testing.T, ctx context.Context, f *forwarder) error {
 		t.Helper()
 
 		var nonRetriableErrSeen bool
 		err := testutils.SucceedsSoonError(func() error {
-			err := f.TransferConnection()
+			err := f.TransferConnection(ctx)
 			if err == nil {
 				return nil
 			}
@@ -2091,7 +2091,7 @@ func TestConnectionMigration(t *testing.T) {
 			require.NoError(t, err)
 
 			// Show that we get alternating SQL pods when we transfer.
-			require.NoError(t, transferConnWithRetries(t, f))
+			require.NoError(t, transferConnWithRetries(t, tCtx, f))
 			require.Equal(t, int64(1), f.metrics.ConnMigrationSuccessCount.Count())
 			require.Equal(t, tenant2.SQLAddr(), queryAddr(tCtx, t, db))
 
@@ -2102,7 +2102,7 @@ func TestConnectionMigration(t *testing.T) {
 			_, err = db.Exec("SET application_name = 'bar'")
 			require.NoError(t, err)
 
-			require.NoError(t, transferConnWithRetries(t, f))
+			require.NoError(t, transferConnWithRetries(t, tCtx, f))
 			require.Equal(t, int64(2), f.metrics.ConnMigrationSuccessCount.Count())
 			require.Equal(t, tenant1.SQLAddr(), queryAddr(tCtx, t, db))
 
@@ -2120,14 +2120,14 @@ func TestConnectionMigration(t *testing.T) {
 			go func() {
 				defer wg.Done()
 				for subCtx.Err() == nil {
-					_ = f.TransferConnection()
+					_ = f.TransferConnection(tCtx)
 					time.Sleep(100 * time.Millisecond)
 				}
 			}()
 
 			// This loop will run approximately 5 seconds.
 			var tenant1Addr, tenant2Addr int
-			for i := 0; i < 100; i++ {
+			for range 100 {
 				addr := queryAddr(tCtx, t, db)
 				if addr == tenant1.SQLAddr() {
 					tenant1Addr++
@@ -2167,7 +2167,7 @@ func TestConnectionMigration(t *testing.T) {
 			err = crdb.ExecuteTx(tCtx, db, nil /* txopts */, func(tx *gosql.Tx) error {
 				// Run multiple times to ensure that connection isn't closed.
 				for i := 0; i < 5; {
-					err := f.TransferConnection()
+					err := f.TransferConnection(tCtx)
 					if err == nil {
 						return errors.New("no error")
 					}
@@ -2201,7 +2201,7 @@ func TestConnectionMigration(t *testing.T) {
 			require.Equal(t, int64(0), f.metrics.ConnMigrationErrorFatalCount.Count())
 
 			// Once the transaction is closed, transfers should work.
-			require.NoError(t, transferConnWithRetries(t, f))
+			require.NoError(t, transferConnWithRetries(t, tCtx, f))
 			require.NotEqual(t, initAddr, queryAddr(tCtx, t, db))
 			require.Nil(t, f.ctx.Err())
 			require.Equal(t, initSuccessCount+1, f.metrics.ConnMigrationSuccessCount.Count())
@@ -2223,7 +2223,7 @@ func TestConnectionMigration(t *testing.T) {
 			lookupAddrDelayDuration = 10 * time.Second
 			defer testutils.TestingHook(&defaultTransferTimeout, 3*time.Second)()
 
-			err := f.TransferConnection()
+			err := f.TransferConnection(tCtx)
 			require.Error(t, err)
 			require.Regexp(t, "injected delays", err.Error())
 			require.Equal(t, initAddr, queryAddr(tCtx, t, db))
@@ -2319,7 +2319,7 @@ func TestConnectionMigration(t *testing.T) {
 		time.Sleep(2 * time.Second)
 		// This should be an error because the transfer timed out. Connection
 		// should automatically be closed.
-		require.Error(t, f.TransferConnection())
+		require.Error(t, f.TransferConnection(tCtx))
 
 		select {
 		case <-time.After(10 * time.Second):

